collector/collector.rs

```diff
@@ -0,0 +1,468 @@
+                    _ => false,
+                };
+                (true, Some(baker_id), committee_status, finalization_committee)
+            }
+        },
+    })
+}
+
+// Helper functions and helper traits
+
+/// Converts a GRPC timestamp to a string
+fn from_unix(timestamp: grpc::Timestamp) -> Result<String> {
+    let unix_time: i64 = timestamp.value.try_into()?;
+    match chrono::Utc.timestamp_millis_opt(unix_time).single() {
+        Some(t) => Ok(t.to_rfc3339_opts(SecondsFormat::Nanos, true)),
+        None => bail!("Invalid timestamp received from Node"),
+    }
+}
+
+/// Gets a BlockHashInput from an optional blockhash option. Will panic if input
+/// is None
+fn get_block_hash_input(hash_opt: Option<grpc::BlockHash>) -> Result<grpc::BlockHashInput> {
+    let hash = hash_opt.ok_or(anyhow!("Invalid hash received from Node"))?;
+    Ok(grpc::BlockHashInput {
+        block_hash_input: Some(grpc::block_hash_input::BlockHashInput::Given(hash)),
+    })
+}
+
+/// A checked division. Returns None on division by zero.
+fn checked_div(a: f64, b: f64) -> Option<f64> {
+    let div = a / b;
+    if !div.is_nan() {
+        Some(div)
+    } else {
+        None
+    }
+}
+
+/// Turns a blockhash into a hex-encoded string
+fn hash_to_hex(x: Option<grpc::BlockHash>) -> Result<String> {
+    Ok(hex::encode(x.req()?.value.clone()))
+}
+
+/// Defining a "Require trait". Helps make sure fields are present, otherwise
+/// throws error
+trait Require {
+    type A;
+    fn req(&self) -> Result<&Self::A, anyhow::Error>;
+}
+
+impl<A> Require for Option<A> {
+    type A = A;
+
+    fn req(&self) -> Result<&Self::A> {
+        match self {
+            Some(v) => Ok(v),
+            None => bail!("Missing field in response from Node"),
+        }
+    }
+}
+
+/// Parsing user NodeName input.
+#[derive(Clone, Debug)]
+struct NodeName(Vec<String>);
+
+impl FromStr for NodeName {
+    type Err = anyhow::Error;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        Ok(Self(input.split_whitespace().map(ToOwned::to_owned).collect::<Vec<String>>()))
+    }
+}
+
+impl fmt::Display for NodeName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", &self.0[..].join(" "))
+    }
+}
+
+// Loggers
+
+/// Sets up a logger that logs to stderr.
+pub fn setup_logger(trace: bool, debug: bool, no_log_timestamp: bool) {
+    let env = if trace {
+        Env::default().filter_or("LOG_LEVEL", "trace")
+    } else if debug {
+        Env::default().filter_or("LOG_LEVEL", "debug")
+    } else {
+        Env::default().filter_or("LOG_LEVEL", "info")
+    };
+
+    let mut log_builder = Builder::from_env(env);
+    if no_log_timestamp {
+        log_builder.format_timestamp(None);
+    } else {
+        log_builder.format(|buf, record| {
+            writeln!(buf, "{}: {}: {}", buf.timestamp_nanos(), record.level(), record.args())
+        });
+    }
+    log_builder.filter(Some("tokio_reactor"), LevelFilter::Error);
```

Comment:
These loggers are copied from the Node.

Most of these filters don't seem to be necessary, will look deeper into it tomorrow.